In [ ]:
!pip install underthesea py_vncorenlp ftfy emot vinorm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.9/20.9 MB 31.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.4/54.4 kB 4.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.5/61.5 kB 6.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.4/40.4 MB 14.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 32.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 657.8/657.8 kB 21.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 22.4 MB/s eta 0:00:00
  Created wheel for py_vncorenlp: filename=py_vncorenlp-0.1.4-py3-none-any.whl size=4306 sha256=033db953e14d5f01d4b545f74f51ed9a73e36596854c8b8ee6507f8e015c1632
  Stored in directory: /root/.cache/pip/wheels/d5/d9/bf/62632cdb007c702a0664091e92a0bb1f18a2fcecbe962d9827
Successfully built py_vncorenlp


In [ ]:
import emot
emot_obj = emot.core.emot()

In [ ]:
import py_vncorenlp

py_vncorenlp.download_model(save_dir='/content')
rdrsegmenter = py_vncorenlp.VnCoreNLP(annotators=["wseg"], save_dir='/content')

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
from os import path

root_path = "/content/gdrive/Shareddrives/BA/R Python/Đồ án/src/"
dataset_path = path.join(root_path, "dataset")
annotated_path = path.join(dataset_path, 'annotated')

!ls "$dataset_path"
!ls "$annotated_path"

annotated  raw	val_annotated
annotatd-5.xlsx     annotated_data.xlsx  val_annotated_data_1500.csv
annotated_data.csv  data.xlsx		 val_annotated_data_1500.xlsx


In [ ]:
import pickle as pkl

with open(path.join(raw_path, 'fashion.p'), 'rb') as f:
  fashion_data = pkl.load(f)

rating_star        cmtid       itemid  \
0                5  12486152255  23676957049   
1                5  12999867370  23754591984   
2                4  12360238836  23676957049   
3                5  12877630757  23182460672   
4                1  12154882262  16694691238   
...            ...          ...          ...   
14298            1  12198892833  22313943551   
14299            2  13257460773  19093790478   
14300            1  13061370615   7141012796   
14301            5  13194167688  21695044273   
14302            4  10421157301   6888422353   

                                                    name  \
0       Sơ mi trễ vai tay ngắn, sơ mi nữ kiểu dáng di...   
1       Quần jeans rách ống suông, quần jeans rách ốn...   
2       Sơ mi trễ vai tay ngắn, sơ mi nữ kiểu dáng di...   
3       Áo thun nữ dài tay dây rút 2 bên, cổ tròn, fo...   
4       ÁO THUN UNISEX/ÁO THUN MẪU CHÂU ÂU RAGE OF TH...   
...                                                  ...   
14298   💛 Loại Đẹp 💛 Tất thể thao cổ cao ngắn nam nữ ...   
14299   Keith House siêu dễ thương dáng rộng phong cá...   
14300         Áo thun cổ tròn lệch vai nhiều màu Ulzzang   
14301   Áo vest len cổ chữ V dáng rộng dành cho nữ Áo...   
14302   Quần Cullotes vải nữ Quần ống rộng lưng cao Q...   

                                                 comment in - out star  
0       Tuyệt vời, giao hàng nhanh, giá rẻ, đóng gói ...       in    5  
1       Đẹp quá, màu sắc thật tuyệt vời. Còn ngần ngạ...       in    5  
2       Đẹp đấy nhưng hơi rộng nên mình phải để ở phí...       in    5  
3       "Xinh lắm mọi người. Chất vải mềm mịn, form d...       in    5  
4                         Xấu, tốt nhất là không nên mua       in    1  
...                                                  ...      ...  ...  
14298                               24k 1 đôi nhé mẹ =))      out    0  
14299                                150k áo chất quá tệ       in    1  
14300  🥲Hình màu đen nhưng màu xanh than. Form áo đẹp...       in    3  
14301   😎Bao bì rất đẹp. Đúng như dự đoán, người bán ...       in    5  
14302  🔥🔥🔥Tặng 100 khách hàng giá gốc ngay hôm nay!!!...      out    0  

[14303 rows x 7 columns]

In [ ]:
# @title data_df -> drop dup -> drop empty -> clean_df
dup_df = data_df.drop_duplicates(subset = ['comment'])
clean_df = dup_df.drop(dup_df[dup_df['comment'] == ''].index)

print("DUPLICATE :", len(data_df) - len(data_df.drop_duplicates(subset = ['comment'])))
print("EMPTY     :", len(data_df) - len(data_df.drop(data_df[data_df['comment'] == ''].index)))
print("FINAL     :", len(data_df) - len(clean_df))

clean_df = clean_df.reset_index(drop = True)
clean_df

DUPLICATE : 4088
EMPTY     : 3356
FINAL     : 4089


cmtid       itemid  \
0      13239687733  23886482112   
1      13239626315  23886482112   
2      13239565011  23886482112   
3      13175587025  18894188708   
4      13253091864  18894188708   
...            ...          ...   
14299  12768980268  22648063543   
14300  13249750307  22648063543   
14301  13239621971  22386234995   
14302  13240092618  22386234995   
14303  13339211202  22386234995   

                                                    name  \
0                    Áo len cardigan trơn dáng rộng AO26   
1                    Áo len cardigan trơn dáng rộng AO26   
2                    Áo len cardigan trơn dáng rộng AO26   
3      Áo Khoác Thom , Áo Khoác Lông Cừu 4 Sọc Cánh T...   
4      Áo Khoác Thom , Áo Khoác Lông Cừu 4 Sọc Cánh T...   
...                                                  ...   
14299  Quần kaki túi hộp nữ form suông rộng MIAA lưng...   
14300  Quần kaki túi hộp nữ form suông rộng MIAA lưng...   
14301  N07 Sét Nhung QC Phối Viền, Sét Bộ Ngủ Cho Nữ ...   
14302  N07 Sét Nhung QC Phối Viền, Sét Bộ Ngủ Cho Nữ ...   
14303  N07 Sét Nhung QC Phối Viền, Sét Bộ Ngủ Cho Nữ ...   

                                                 comment  rating_star  
0      mặc khá mát vải này cũng ổn đẹp tuyệt với cho ...            5  
1      hàng giao nhanhh chất lươngg sản phẩm tuyệt vo...            5  
2      Sp rất đẹp ,thơm, shop giao siêu nhanh, chất v...            5  
3      Màu sắc:ko\nChất liệu:ko cos\nĐúng với mô tả:k...            1  
4      Màu sắc:đen\nĐúng với mô tả:cũng OK\nChất liệu...            1  
...                                                  ...          ...  
14299  Trời ơi quần đẹp xỉu tụi bây ơi, thề mua được ...            5  
14300  Quần vải đẹp, màu xinh lắm, form quần ưng nha,...            5  
14301  M63 45kg mặc sz L vừa chiều dài nha. Hàng đẹp ...            5  
14302  "Chất liệu như mô tả\nNhìn đẹp, sang nên mua đ...            5  
14303  Màu sắc:đen\nĐúng với mô tả:đúng\nChất liệu:nh...            5  

[14304 rows x 5 columns]

Sau khi quan sát nhóm nhận thấy có những comment bị lặp lại hoặc những comment bị rỗng **(lý do)** nên sẽ thực hiện xoá những dòng dữ liệu này:

1. dup_df = data_df.drop_duplicates(subset=['comment']): Loại bỏ các dòng trùng lặp dựa trên cột 'comment'. Kết quả là dup_df sẽ là một DataFrame mới không chứa các bình luận trùng lặp.

2. clean_df = dup_df.drop(dup_df[dup_df['comment'] == ''].index): Loại bỏ các dòng có giá trị rỗng trong cột 'comment'. Nếu một dòng có 'comment' rỗng, nó sẽ được loại bỏ khỏi DataFrame dup_df. Kết quả là clean_df là DataFrame cuối cùng sau khi loại bỏ các bình luận trùng lặp và các bình luận rỗng.

3. clean_df = clean_df.reset_index(drop=True): Reset index của DataFrame clean_df để có index liên tục từ 0 đến len(clean_df) - 1.

**Tổng số các dòng bị loại bỏ vì bị trùng lặp và bị rỗng là 4088 dòng, do đó bộ dữ liẹu còn lại 14304 dòng.**


In [ ]:
tests = clean_df['comment'][:100].values.tolist()
tests

['mặc khá mát vải này cũng ổn đẹp tuyệt với cho 10 điêm',
 'hàng giao nhanhh chất lươngg sản phẩm tuyệt voiii nen mua nhaa',
 'Sp rất đẹp ,thơm, shop giao siêu nhanh, chất vải xịn dã man',
 'Màu sắc:ko\nChất liệu:ko cos\nĐúng với mô tả:ko\n\nChất lượng sản phẩm kém \nK\nNhu cầu mua lại không bao giờ\nĐộ lừa đảo quá cao nhé mng',
 'Màu sắc:đen\nĐúng với mô tả:cũng OK\nChất liệu:chưa thấy\n\nVì rẻ lên đánh giá 5* vì kg ở nhà nhưng vẫn cảm ơn shop 45k. Có Đức quá',
 'Đúng với mô tả:mua ao giao. Cái khác\n\nNhắn tin shop k tl \nTệ\nShop k uy tín',
 'Đúng với mô tả:không đúng\n\nĐặt áo lông giao quần trẻ con :)) đã thế quần còn bẩn và cũ',
 'Đúng với mô tả:không\nChất liệu:không đúng với đơn hàng đã đặt\nMàu sắc:trắng nâu đen\n\nRõ ràng đặt 1 lần 5 áo nhưng giao về chỉ có 1 áo\nĐã vậy còn sai mẫu không giống với mẫu mình đặt \nYêu cầu trả hàng\nBao giờ hoàn trả thành công thì mình sẽ đánh giá lại 5 sao',
 'Màu sắc:sai hàng\nĐúng với mô tả:sai hàng\nChất liệu:sai hàng\n\nShop làm ăn kiểu đé*

In [ ]:
!git clone https://github.com/v-nhandt21/Vinorm.git
!ls

Cloning into 'Vinorm'...
remote: Enumerating objects: 125, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 125 (delta 4), reused 3 (delta 3), pack-reused 120
Receiving objects: 100% (125/125), 202.66 MiB | 31.24 MiB/s, done.
Resolving deltas: 100% (32/32), done.
Updating files: 100% (128/128), done.
gdrive	models	sample_data  Vinorm  VnCoreNLP-1.2.jar


In [ ]:
import os

slangs_dict = {}
for file in os.listdir('Vinorm/vinorm/Mapping'):
  if file not in ['Teencode.txt', 'Number.txt']:
    continue

  print(file, ':', end = ' ')
  file_path = 'Vinorm/vinorm/Mapping/' + file

  with open(file_path, 'r', encoding = 'utf8') as f:
    data_slangs = f.readlines()
    print(len(data_slangs))

    for i in data_slangs:
      i = i.replace("\n", "").lower().strip()
      i = i.split("#")

      for punc in ['(', ')', '.', '+', '*']:
        i[0] = i[0].replace(punc, '\\' + punc)

      slangs_dict[i[0]] = i[1]

slangs_dict.update({
    'sp': 'sản phẩm',
    'tl': 'trả lời',
    'mn': 'mọi người',
    'mng': 'mọi người',
    'mngf': 'mọi người',
    'h': 'giờ',
    'v': 'vậy',
    'cgi': 'cái gì',
    'vs': 'với',
    'sv': 'sinh viên',
    'sop': 'shop',
    'lmj': 'làm gì',
    'e': 'em',
    'dc': 'được',
    'nx': 'nữa',
    'kg': 'không',
    'bt': 'biết',
    'l': 'làm',
    'ntin': 'nhắn tin',
    'j': 'gì',
    'v': 'vậy',
    'lm': 'làm',
    't': 'tao',
    's': 'sao',
    'ngta': 'người ta',
    'z': 'vậy',
    'nt': 'nhắn tin',
    'ib': 'inbox',
    'lh': 'liên hệ',
    'kh': 'không'
})

len(slangs_dict)

Teencode.txt : 484
Number.txt : 10


521

Nhóm xây dựng một từ điển slangs_dict chứa những từ viết tắt và teencode để sử dụng cho việc thay thế những từ ngữ viết tắt và teencode này chuyển thành từ ngữ chuẩn. Bên cạnh đó, khi thực hiện gán dữ liệu nhóm có thực hiện lưu lại những từ ngữ viết tắt khác và cập nhật vào từ điển slangs_dict bằng cách update từ điển này thường xuyên.

**Sau khi thực hiện nhóm thu được một từ điển từ ngữ viết tắt và teencode gồm : 521 từ**



In [ ]:
import regex as re

with open(path.join(root_path, 'slangs update.txt'), 'r', encoding = 'utf8') as f:
  data = f.readlines()
  data = [re.sub(r'[,‘’\';]', '', i).lower() for i in data]
  data = [i.replace('\n', '').replace('\ufeff', '') for i in data]
  new_slang = {}

  for i in data:
    j = i.split(":")
    new_slang[j[0].strip()] = j[1].strip()
slangs_dict.update(new_slang)
len(slangs_dict)

606

Đoạn trên nhóm thực hiện việc cập nhật từ điển slangs_dict bằng cách đọc dữ liệu từ một tệp tin văn bản 'slangs update.txt' và thêm vào các ánh xạ mới.

**Từ điển về từ ngữ viết tắt và teencode cuối cùng của nhóm là 606 từ**

In [ ]:
VIETNAMESE_TEXT = 'àáãạảăắằẳẵặâấầẩẫậèéẹẻẽêềếểễệđìíĩỉịòóõọỏôốồổỗộơớờởỡợùúũụủưứừửữựỳỵỷỹýÀÁÃẠẢĂẮẰẲẴẶÂẤẦẨẪẬÈÉẸẺẼÊỀẾỂỄỆĐÌÍĨỈỊÒÓÕỌỎÔỐỒỔỖỘƠỚỜỞỠỢÙÚŨỤỦƯỨỪỬỮỰỲỴỶỸ'
LATIN_VI = r'[A-Za-z' + VIETNAMESE_TEXT + r']'

regex_dict = {
    r'(?<=\d+\s)k\b': 'nghìn',
    r'(?<=\d+\s)m(?=\s\d+)\b': 'mét',
    r'(?<=\d+\s)kg\b': 'kilogram',
    r'(?<=size\s)s\b': 'nhỏ',
    r'(?<=size\s)m\b': 'trung bình',
    r'(?<=size\s)l\b': 'lớn',
    r'(?<=size\s)xl\b': 'rất lớn',
    r'(?<=size\s)xxl\b': 'rất rất lớn',
}

Do tính chất đặc biệt của ngành hàng Thời trang, các comment có thể chứa thông tin về kích thước, chiều cao, cân nặng của khách hàng dưới dạng viết tắt như S, M, L. Đế trách việc chuyển đổi từ viết tắt bị sai, nhóm tiếng hành xây dụng từ điển cho các nhóm từ này.

In [ ]:
import underthesea as u
import ftfy
import string
from vinorm import TTSnorm
import unicodedata

def demoji(text):

  text = text.strip()
  text = ' ' + text + ' '

  emojis = emot_obj.emoji(text)
  emoticons = emot_obj.emoticons(text)
  for k, v in zip(emojis['value'] + emoticons['value'], emojis['mean'] + emoticons['mean']):

    for punc in ['(', ')', '.', '+', '*']:
      k = k.replace(punc, '\\' + punc)

    v = v.replace(':', '')
    v = '<emoji>' + v + '</emoji>'

    pattern = r'\s' + k + r'\s'
    text = re.sub(pattern, ' ' + v + ' ', text)

  text = text.lower()
  text = text.strip()

  return text

def pre_processing(text):

  text = text.lower()
  text = unicodedata.normalize('NFC', text)

  # Clean space
  text = text.replace("\n", " ")
  text = re.sub(r'\s+', ' ', text)

  # Fix encoding, punc
  text = text.strip()
  text = ftfy.fix_encoding(text)
  text = ftfy.fix_text(text)
  text = u.text_normalize(text)

  # Turn emoji, emoticon to text
  text = demoji(text)

  # Fix dup punc
  text = re.sub(r'([!"#$%&\'()*+,\-\.\/:;<=>?@[\\\]^_`{|}~])(?:\s*\1)+', r' \1 ', text)
  text = re.sub(r'\s+', ' ', text)

  # Remove punc
  text = re.sub(r'[!"#$%&\'()*+,\-\.:;=?@[\\\]^_`{|}~]', ' ', text)
  text = re.sub(r'<(?!\/?[^\s]+>)|(?<!<\/?[^\s]+)>|(?<!<)\/(?![^\s]+>)', ' ', text)
  text = re.sub(r'\s+', ' ', text)

  # Normalize with regex
  text = re.sub(r'((?<!' + LATIN_VI + r'\s?)' + LATIN_VI + r'{1,3}(?!' + LATIN_VI + r'\s?))', r' \1 ', text)
  text = text.strip()
  text = re.sub(r'\s+', ' ', text)
  for pattern, value in regex_dict.items():
    text = re.sub(pattern, value, text)

  # Fix slang
  text = " ".join([slangs_dict.get(word, word) for word in text.split()])

  # Word segment
  try:
    text = rdrsegmenter.word_segment(text)[0]
  except:
    pass

  # Fix token
  text = re.sub(r'<\s?(\/)?\s([^\s]+)\s>', r'<\1\2>', text)

  text = re.sub(r'\s+', ' ', text)
  return text.strip()

for test in tests:
  print(pre_processing(test))
  print()

print("=" * 50)


mặc khá mát vải này cũng ổn đẹp tuyệt với cho 10 điêm

hàng giao nhanhh chất lươngg sản_phẩm tuyệt voiii nen mua nhaa

sản_phẩm rất đẹp thơm shop giao siêu nhanh chất vải xịn dã_man

màu_sắc không chất_liệu không cos đúng với mô_tả không chất_lượng sản_phẩm kém không nhu_cầu mua lại không bao_giờ độ lừa_đảo quá cao nhé mọi người

màu sắc đen đúng với mô_tả cũng ok chất_liệu chưa thấy vì rẻ lên đánh_giá năm vì không ở nhà nhưng vẫn cảm_ơn shop 45 nghìn có đức quá

đúng với mô_tả mua ao giao cái khác nhắn_tin shop không trả_lời tệ shop không uy_tín

đúng với mô_tả không đúng đặt áo lông giao quần trẻ_con <emoji> very happy face or smiley </emoji> đã thế quần còn bẩn và cũ

đúng với mô_tả không chất_liệu không đúng với đơn hàng đã đặt màu_sắc trắng nâu đen rõ_ràng đặt một lần năm áo nhưng giao về chỉ có một áo đã vậy còn sai mẫu không giống với mẫu mình đặt yêu_cầu trả hàng bao_giờ hoàn_trả thành_công thì mình sẽ đánh_giá lại năm sao

màu_sắc sai hàng đúng với mô_tả sai hàng chất_liệu sai

'tôi cao một mét 50 nặng 35 kilogram bộ áo 45 nghìn này có vừa không'

Nhóm sử dụng hàm demoji để chuyển đổi emoji và emoticon thành văn bản. Và sử dụng hàm pre_processing để tiền xử lý nội dung của comment bao gồm:
- Chuyển đổi toàn bộ văn bản thành chữ thường
- Chuẩn hóa unicode để đảm bảo rằng các ký tự được biểu diễn theo một cách duy nhất
- Thay thế ký tự xuống dòng bằng khoảng trắng
- Loại bỏ khoảng trắng thừa
- Sửa các lỗi chính tả và biểu hiện văn bản.
- Chuẩn hóa văn bản sử dụng underthesea
- Loại bỏ dấu câu và các ký tự đặc biệt khác
- Sửa các từ viết tắt bằng cách thay thế chúng theo từ điển slangs_dict
- Sử dụng rdrsegmenter để tách từ
- Sửa các token được tạo từ thẻ HTML.

Cuối cùng hàm sẽ trả về văn bản đã được xử lý.

In [ ]:
clean_df['clean_cmt'] = clean_df['comment'].apply(pre_processing)
clean_df = clean_df.drop(clean_df[clean_df['clean_cmt'] == ''].index)
clean_df = clean_df.reset_index(drop = True)

Issue with internal pattern finding emoji: '🇻'
Issue with internal pattern finding emoji: '🇬'


In [ ]:
clean_df

cmtid       itemid  \
0      13239687733  23886482112   
1      13239626315  23886482112   
2      13239565011  23886482112   
3      13175587025  18894188708   
4      13253091864  18894188708   
...            ...          ...   
14298  12768980268  22648063543   
14299  13249750307  22648063543   
14300  13239621971  22386234995   
14301  13240092618  22386234995   
14302  13339211202  22386234995   

                                                    name  \
0                    Áo len cardigan trơn dáng rộng AO26   
1                    Áo len cardigan trơn dáng rộng AO26   
2                    Áo len cardigan trơn dáng rộng AO26   
3      Áo Khoác Thom , Áo Khoác Lông Cừu 4 Sọc Cánh T...   
4      Áo Khoác Thom , Áo Khoác Lông Cừu 4 Sọc Cánh T...   
...                                                  ...   
14298  Quần kaki túi hộp nữ form suông rộng MIAA lưng...   
14299  Quần kaki túi hộp nữ form suông rộng MIAA lưng...   
14300  N07 Sét Nhung QC Phối Viền, Sét Bộ Ngủ Cho Nữ ...   
14301  N07 Sét Nhung QC Phối Viền, Sét Bộ Ngủ Cho Nữ ...   
14302  N07 Sét Nhung QC Phối Viền, Sét Bộ Ngủ Cho Nữ ...   

                                                 comment  rating_star  \
0      mặc khá mát vải này cũng ổn đẹp tuyệt với cho ...            5   
1      hàng giao nhanhh chất lươngg sản phẩm tuyệt vo...            5   
2      Sp rất đẹp ,thơm, shop giao siêu nhanh, chất v...            5   
3      Màu sắc:ko\nChất liệu:ko cos\nĐúng với mô tả:k...            1   
4      Màu sắc:đen\nĐúng với mô tả:cũng OK\nChất liệu...            1   
...                                                  ...          ...   
14298  Trời ơi quần đẹp xỉu tụi bây ơi, thề mua được ...            5   
14299  Quần vải đẹp, màu xinh lắm, form quần ưng nha,...            5   
14300  M63 45kg mặc sz L vừa chiều dài nha. Hàng đẹp ...            5   
14301  "Chất liệu như mô tả\nNhìn đẹp, sang nên mua đ...            5   
14302  Màu sắc:đen\nĐúng với mô tả:đúng\nChất liệu:nh...            5   

                                               clean_cmt  
0      mặc khá mát vải này cũng ổn đẹp tuyệt với cho ...  
1      hàng giao nhanhh chất lươngg sản_phẩm tuyệt vo...  
2      sản_phẩm rất đẹp thơm shop giao siêu nhanh chấ...  
3      màu_sắc không chất_liệu không cos đúng với mô_...  
4      màu sắc đen đúng với mô_tả cũng ok chất_liệu c...  
...                                                  ...  
14298  trời_ơi quần đẹp xỉu tụi bây ơi thề mua được c...  
14299  quần vải đẹp màu xinh lắm form quần ưng nha gi...  
14300  m 63 45 kilogram mặc size làm vừa chiều dài nh...  
14301  chất_liệu như mô_tả nhìn đẹp sang nên mua để m...  
14302  màu sắc đen đúng với mô_tả đúng chất_liệu nhun...  

[14303 rows x 6 columns]

In [ ]:
clean_df.to_csv(path.join(preprocessing_path, 'fashion_v2.csv'), index = False, header = True, encoding = 'utf8')